### ISSUE TITLE
Updating belongsToMany references in embedded document doesn't save both ways

### ISSUE THREAD

**Issue by IamRobin** [jenssegers/laravel-mongodb#527]
I've been struggling with this for a few days now. I've posted this on stack overflow as well, but no reaction so far. This might or might not be an actual issue with the package.

I'm using Laravel 5 with MongoDB with this package. I'm using embedded documents for a certain database structure. For example, I have a Product model which embedsMany objects of a ProductVariant model. This would look like this in the database:

```
[
  {
    "_id": "MongoID",
    "product_name": "T-shirt",
    "description": "Short description",
    "brand": "nike",
    "series": "whatever",
    "is_published": false,
    "slug": "t-shirt",
    "category_ids": [
        "MongoID"
    ],
    "variants": [
    {
      "product_sku": "1289938",
      "description": "Short description",
      "price": "430",
      "slug": "xn1289938-t-shirt",
      "discountPrice": "0",
      "is_published": false
      "_id": {
         "$id": "MongoID"
       }
    }
   ]
 }
]
```

The 'variants' in this document are objects of ProductVariant. In my code of the Product model I'm using:

```
public function variants () {

    return $this->embedsMany('\ProductVariant');

}
```

This works fine. Now I have a different model called Facet, which would add an array of attributes to a ProductVariant. So in my ProductVariant I have:

```
public function facets(){

     return $this->belongsToMany('\Facet');

}
```

Facets are reusable across multiple ProductVariants so that is why I'm trying to save them as a referenced document. However, when I call a sync method for updating the reference with a list of ids...

```
 $productVariant->facets()->sync(Request::json('facets'));
```

...nothing happens to the ProductVariant in the database. However the Facet model gets a correct update in the database:

```
 "_id": "blue_color",
"description": "Blue color",
"type": "color",
"value": "blue",
"product_variant_ids": [
     "MongoID"
 ]
```

In other situations (without the embedded document) I get an updated reference with 'related_ids' on both sides. Also when I don't embed these same models, but save them as a flat document the sync works without any changes to my code. The sync simply doesn't update embedded documents, but only updates the other side of the relationship which is a simple document. Am I doing it wrong or is this operation not supported?

**Comment by hackel** [comment]
I've been struggling with this issue as well, and unfortunately it seems that it is not supported.  I think that it requires some thought, since "product_variant_ids" in the Facet document is completely meaningless, since it doesn't convey which Product they are in, and those IDs could theoretically be duplicated across products.

This could be done by storing embedded documents that are essentially a reverse of the dependency tree:
```
product_variant_ids: [
  {
    _id: MongoID,
    product_id: MongoID
  }
], ...etc.
```
It could keep burying down for multiple levels until it got to the parent object.

My problem with this is, it seems like overkill, and I can't even imagine how nasty the Mongo query would have to be to do anything useful with this data.  The problem seems to be that laravel-mongodb refuses to return a related object unless the ID is present in *both* tables.  A single table should be sufficient.  Yes, you won't be able to get a list of Products from a Facet document, but I can live with that.  I can manually add Facet::hasMany(Products) if I need that information.

This limitation almost makes me think that using a pivot table is actually an easier approach, even though this isn't very no-sql-like  It would also allow storing extra data on the pivot table, which would be extremely useful for a project I'm working on now.  So basically I think some kind of a flag would be nice, that allowed you to toggle the "embedded pivot table" behaviour on a per-collection basis.

Would appreciate any thoughts or alternative approaches.

**Comment by IamRobin** [comment]
Well, it's been over a month since I've asked this question and so far the author hasn't responded, so I guess this is working as he intended it. But I still had to solve my issue. So I created my own BelongsToMany relation that extends the original one and applied it to the embedded model. I created a sync method for that like this:

    public function sync($ids, $detaching = true)
    {
        $current = $this->parent{$this->otherKey} ?: array();

        if($current instanceof Collection) $current = $current->modelKeys();

        $detach = array_diff($current, $ids);

        foreach ( $detach as $id ) {
            $this->related
                 ->where('_id', $id)
                 ->pull($this->foreignKey, $this->parent->_id);
        }

        $attach = array_diff($ids, $current);

        foreach ( $attach as $id ){

            $this->related
                ->where('_id', $id)
                ->push($this->foreignKey, $this->parent->_id);
        }

        $this->parent->{$this->otherKey} = $ids;

        return;
    }

It's probably not the right way. I suspect this increases the number of database transactions needed on write, but it's the only way I got it to work. I simply used the push and pull methods to update arrays. The original sync method didn't seem capable of updating the embedded document. I still don't know why.

Personally, I wouldn't like the approach of having something like a pivot document. I would like to retrieve as much information that I need in one place with a single read. Also, it doesn't hurt to have product_ids on a facet, because when you delete that facet you also need to remove it from the products it is attached to. In this case I would call the sync with an empty array before deleting the facet. 

I don't know if I'm doing this right, so feel free to correct me, but for now it saved my case.
